### ISSUE TITLE
`FOAM_USER_LIBBIN` not properly configured?

### ISSUE THREAD

**Issue by petebachant** [gerlero/openfoam-app#235]
I am trying to use [turbinesFoam](https://github.com/turbinesFoam/turbinesFoam) in this environment. It compiles fine, but I'm not able to load the library, apparently because the user lib directory is not being searched at runtime:

```
--> FOAM Warning : 
    From void *Foam::dlLibraryTable::openLibrary(const fileName &, bool)
    in file db/dynamicLibrary/dlLibraryTable/dlLibraryTable.C at line 188
    Could not load "turbinesFoam"
dlopen(libturbinesFoam.dylib, 0x0009): tried: '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/openmpi/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/usr/opt/fftw/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/usr/opt/mpfr/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/usr/opt/gmp/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/usr/opt/cgal/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/usr/opt/boost/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/usr/opt/open-mpi/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/dummy/libturbinesFoam.dylib' (no such file), 'libturbinesFoam.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/openmpi/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/dummy/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/openmpi/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/libturbinesFoam.dylib' (no such file), '/Volumes/OpenFOAM-v2312/platforms/darwin64ClangDPInt32Opt/lib/dummy/libturbinesFoam.dylib' (no such file), '/usr/lib/libturbinesFoam.dylib' (no such file, not in dyld cache), 'libturbinesFoam.dylib' (no such file)
```

```sh
$ echo $FOAM_USER_LIBBIN/
/Users/pete/OpenFOAM/pete-v2312/platforms/darwin64ClangDPInt32Opt/lib/
```

Is there some additional configuration that needs to be done? The `porousMicroTransport` example given in the README seems to put the libraries into `FOAM_MODULE_LIBBIN`, but that variable is not defined in my environment.

**Comment by gerlero** [comment]
Yeah, that's a regression in OpenFOAM v2312 introduced by the `rpath` changes in https://develop.openfoam.com/Development/openfoam/-/issues/2948. I admit I have seen it too but haven't reported it upstream. It indeed causes user binaries to no longer carry references to user libraries, leading to the error you describe.

A workaround for this is to use a launch script that begins by sourcing `$WM_PROJECT_DIR/bin/tools/RunFunctions` ([e.g.](https://github.com/gerlero/porousMicroTransport/blob/753963fd3fced6ad1c285817277bb7d1683c471d/tutorials/moistureDiffusivityFoam/validity/base/run)); this will add the user lib directory to the dynamic search paths. Unfortunately, installing to `FOAM_MODULE_LIBBIN` shouldn't help as that build variable gets translated into the user (or site) lib dir by `wmake` itself, leading to the same problem. Another workaround should be to use OpenFOAM v2306 or earlier, but that's not ideal either.

**Comment by petebachant** [comment]
Interestingly that doesn't seem to work for me, as I was already sourcing those run functions in my script. In your example, you don't list the lib in `controlDict`, which I assume is done because your linking the lib when compiling the solver. In my case, I'm trying to use the lib with `pimpleFoam` to use an `fvOption` inside, but I can't look up the lib at run time.

I appreciate your help nonetheless!

**Comment by gerlero** [comment]
@petebachant If you're able to test, does your use case work with OpenFOAM v2306?

**Comment by petebachant** [comment]
> Well, thanks; it was worth a shot. This means this issue is not a regression from the added `rpath` stuff, but a different (very likely [SIP](https://support.apple.com/en-us/102149)-related) issue.
> 
> I'm running out of ideas of workarounds, but you might still want to try the same launch script interpreted by the app's bundled Bash shell (e.g., by replacing the shebang line in the script with something like `#!/Volumes/OpenFOAM-v2312/usr/opt/bash/bin/bash`; I assume you're using `#!/bin/bash` or similar). This bundled Bash install doesn't enforce SIP, so I guess there's a chance that the necessary library lookup works that way.

No luck unfortunately. If I can figure out another workaround I will report back here. Thanks again.

**Comment by petebachant** [comment]
A bit of a hack, but this appears to work if I run it at the start of a shell session:

```sh
export DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:$FOAM_USER_LIBBIN
```

**Comment by gerlero** [comment]
@petebachant Thanks! Then it's likely [this bug mentioned here](https://develop.openfoam.com/Development/openfoam/-/issues/2948#note_61598). 

I'll wait a few days for a reply on that thread in case it's already fixed or being tracked somewhere. After that we can consider whether it's worth adding that workaround here.

**Comment by mrklein** [comment]
@gerlero I have reported new issue: https://develop.openfoam.com/Development/openfoam/-/issues/3098. The problem is in multi-folder argument to `_foamAddLib` function on line 233 of `etc/config.sh/setup` (or line 201 in case of `etc/config.csh/setup`).

I do not know if `open foam-app` users can modify files on the disk image, if yes, solution 2 from the issue could be easily implemented by a user.

**Comment by gerlero** [comment]
@mrklein Thanks! I'll try to make a backport here of whichever solution makes into OpenFOAM. Users can't alter the disk image so the change would be incorporated into the configuration step.

**Comment by gerlero** [comment]
Thanks again to everyone! This should now be fixed in OpenFOAM.app 1.13.0 (just released).
